### ISSUE TITLE
Limit Proxy Cache Images

### ISSUE THREAD

**Issue by phin1x** [goharbor/harbor#13231]
With the new version of Harbor it can now be used as a proxy cache. As administrator I would like to have the possibility to restrict the images that are pulled (whitelist). This would save us a lot of replication rules.

**Comment by xaleeks** [comment]
can you expand? how does this save a lot of replication rules? @phin1x

**Comment by phin1x** [comment]
If i create a proxy project (for example docker hub), all users with pull permission can pull all images from the proxied registry. if i only want to limit the images, which can be pulled (for example only dockerhub library images should be allowed), i cannot use the proxy cache project type.
in our company, images from third party registries like dockerhub, quay, gcr, gitlab and so on are forbidden by default and images need a security check befor they are allowed to be used. To ensure that, we create a replication rule for every granted image (about 30 rules at the momement, ascending trend).
if we can limit the images, which can be pulled through a proxy cache project, this would save us a lot of replication rules, 
especially if the whitelisting would be based on wildcard filters.

**Comment by rayisbadat** [comment]
I would just like to add my +1 to this feature request.
 Currently i am syncing about 30+ repos from dockerhub.  But if there was a Name like feature on the proxy cache as there is for the normal replications i could get rid most of those.  Ex: `Name: library/**` or `Name: library/{alpine,centos,ubuntu}` .  It would would allow me to proxy official docker hub images, and prevent random people's personal images.

**Comment by nnsense** [comment]
This feature would allow us to open the proxy to internet without worrying about people using our proxy to circumvent docker hub restrictions.

**Comment by Danielkem** [comment]
I also believe that this feature would be of great value. A common setup would be to host a Harbor instance on a cluster and set policies to only allow pulling images from that Harbor instance for security reasons. However, it would be quite redundant in many cases to also replicate all images that are maintained by some known and trusted publisher. In this case I would like to define a proxy cache to a single Docker Hub publisher/user/project/repository while not opening up the ability to pull images from any public repository that is hosted at Docker Hub.

Is anyone looking at or working on implementing this kind of feature?

**Comment by pafra** [comment]
We would need this feature too!

**Comment by AYDEV-FR** [comment]
+1
I need this feature too.
And this feature can be add with it

https://github.com/goharbor/harbor/issues/13242

**Comment by hri-op** [comment]
+1
We need also this feature.. otherwise we use replication to replicate only the images we want to be present in our local harbor registry.

**Comment by damon-xu-betsson** [comment]
+1
We need this feature as well.

**Comment by gauthiersiri** [comment]
That would be amazing! +1

**Comment by caimez** [comment]
+1
Agreed!
Another idea to add here would be image scanning at time of proxy pull and potential refusal of the image if there are a certain number of low, medium, high vulnerabilities.

**Comment by atchadwick** [comment]
+1
Would love to see this feature, especially with the continued focus on security for images, being able to whitelist a proxy cache is a nice balance of accessibility for users without the need to keep a huge replication rule list or pre pull a bunch of unused images.

**Comment by lukeelten** [comment]
> +1 Would love to see this feature, especially with the continued focus on security for images, being able to whitelist a proxy cache is a nice balance of accessibility for users without the need to keep a huge replication rule list or pre pull a bunch of unused images.

I agree with @atchadwick . This feature would greatly increase our security measures against malicious images.

**Comment by gira0** [comment]
+1

This would definitely be a helpful feature to ensure only trusted sources from docker hub would be available, right now. Every few weeks there's another article about the rise of malicious images on docker hub since it's the defacto default registry.

**Comment by paspflue** [comment]
+1

**Comment by tuxpeople** [comment]
I'de also desperately need this feature for a customer. With only a few projects in place yet, we already have a big mess of replication rules because of the company policies where security needs to allow which images to use. As proxy cache can't be filtered, we have to replicate each and every image, The problem is, that replication rules with stuff like library/{alpine,centos,ubuntu} do also not work, because without tag specification it would sync may to many images. So basically we end up with doing a replication job for every single allowd image with the tags we need. Would be much easier if there's a whitelist for proxy caches.

**Comment by gsforza** [comment]
+1

**Comment by igorrudyk** [comment]
Hi guys,
Is it possible to implement?

**Comment by LucEast** [comment]
This is a big NEED! 
It would be especially crucial for enterprises that are ISO 27001 certified (or more specific "BSI" certfied in germany) where one isn't allowed to pull images directly from the web.

Let's say one runs a rke2 cluster. I would allow to pull all `rancher/**` images trough image proxy cache. Other Images still need to be manually pushed or replicated to harbor.

**Comment by reasonerjt** [comment]
Unfortunately, we are fully occupied in v2.15 timeframe.  We're aligned this is valuable and we'll try to address this in v2.16.

**Comment by reasonerjt** [comment]
Having to kick it out of 2.16 due to bandwidth issue, keeping it in "backlog"

**Comment by AYDEV-FR** [comment]
Hi everyone 👋

I'd like to volunteer to implement this feature. I've recently contributed to the proxy cache area and I'm familiar with the relevant code paths. This feature is also a direct pain point for me: I currently manage over 400 replication rules via Terraform, and a native allowlist filter on proxy cache projects would dramatically simplify that.

### Proposed Design
The core idea is to introduce an image allowlist (repository name patterns + optional tag patterns) configurable per proxy cache project. Any pull request that doesn't match the allowlist would be rejected with a 403 Forbidden before Harbor proxies anything from the upstream registry.

### Data model
Store the filter rules in project_metadata (consistent with how proxy_speed_kb is already stored), or as a dedicated table `proxy_cache_filters` linked to project_id.

Each rule would contain:
- repository_pattern — glob or regex, e.g. library/nginx, myorg/*
- tag_pattern (optional) — e.g. v1.*, latest

### Backend changes
The enforcement point would be in `src/server/middleware/repoproxy/proxy.go`, inside the preCheck() function, right after the project metadata is loaded and before the remote pull is triggered. A new helper `isAllowed(art lib.ArtifactInfo, filters []ProxyCacheFilter)` bool would evaluate the patterns against `art.Repository` and `art.Tag`.
If the allowlist is empty, the current behavior (allow everything) is preserved for full backward compatibility.
API changes
Extend `POST /api/v2.0/projects` and `PUT /api/v2.0/projects/{project_name}` to accept a `proxy_cache_filters` field, similar to how cve_allowlist is handled today.
UI changes
Add a filter configuration section in the proxy cache project creation/edit form (similar to the tag retention rules UI).

Maybe in the future we can add same logic for CVE Image Whitelist ?

I'm happy to start with a design doc / proposal if the maintainers would like to review the approach before I open a PR. @stonezdj would love your input on whether `project_metadata` or a dedicated table is preferred here.

**Comment by reasonerjt** [comment]
@wy65701436 came back from Kubecon EU with the feedback.  

After discussion, we'll de-prioritize some work and prioritize this one in v2.16.0

**Comment by shibaPuppy** [comment]
Is it possible to add label-based proxy cache functionality to the filter?
